genpolicy: add support for cc-managed-csi

sprt opened the pull request:
<!--
COMMENT BLOCKS WILL NOT BE INCLUDED IN THE PR.
Feel free to delete sections of the template which do not apply to your PR, or add additional details
-->

###### Merge Checklist  <!-- REQUIRED -->
<!-- You can set them now ([x]) or set them later using the Github UI -->
<!-- **All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)* -->
- [x] Followed patch format from upstream recommendation: https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#patch-format
  - [x] Included a single commit in a given PR - at least unless there are related commits and each makes sense as a change on its own.
- [x] Aware about the PR to be merged using "create a merge commit" rather than "squash and merge" (or similar)
- [x] genPolicy only: Ensured the tool still builds on Windows
- [x] genPolicy only: Updated sample YAMLs' policy annotations, if applicable
- [x] The `upstream-missing` label (or `upstream-not-needed`) has been set on the PR. 

###### Summary <!-- REQUIRED -->
<!-- Quick explanation of WHAT changed and WHY. -->
Add support to genpolicy for our CC Azure Disk drivers.

###### Test Methodology
<!-- How was this test validated? i.e. local build, pipeline build etc. -->
Manual testing

danmihai1 commented on src/tools/genpolicy/src/mount_and_storage.rs:
```diff
@@ -107,6 +107,23 @@ pub fn get_mount_and_storage(
+                // computed by the shim as follows:
+                //   source = PCI device ID (e.g. "02")
+                //   mount_point = "/run/kata-containers/sandbox/storage/$(base64_encode(source))"
+                source: "".into(),
```
It sems like the source should be similar to "$(pci-device-id)" - similar to the current "$(node-name)" from rules.rego.

Then, either the Agent of rules.rego would decode the incoming pci-device-id and send it to the Rego rules. Not sure how easy is to base64 decode in Rego (including in the upcoming Regorus), so the Agent is typically a better place to decode.

Then, the Rego rules would verify the format of the pci-device-id - e.g., by using some regex.

sprt replied:
Not sure how we'd pass pci-device-id from the agent to Rego? Do we do that with anything else at the moment?

danmihai1 replied:
Example of "decoding" data before sending it to Rego: https://github.com/microsoft/kata-containers/commit/d3181c84dbee2e0388700df21ec61e675f86f775

In that example, the decoding step is:

PathBuf::from(OsStr::from_bytes(&req.data))

danmihai1 commented on src/tools/genpolicy/rules.rego:
```diff
@@ -809,7 +816,9 @@ allow_storages(p_storages, i_storages, bundle_id, sandbox_id) {
     print("allow_storages: p_count =", p_count, "i_count =", i_count)
 
-    p_count == i_count
+    # If it's not a virtio-blk storage, genpolicy creates a superfluous
```
Let's discuss examples - I don't understand why we would end up with superfluous storage objects.

sprt replied:
For managed-csi, there's no storage object created by the shim. For cc-managed-csi, there is one, representing the virtio-blk device.

danmihai1 replied:
Hopefully, the persistentVolumeClaim you used in the change below knows that the user wants cc-managed-csi (and not managed-csi), and then we add the missing block Storage to the policy? Then the new Storage wouldn't be superfluous anymore and the storage would continue to work.

danmihai1 commented on src/tools/genpolicy/src/config_map.rs:
```diff
@@ -124,6 +125,7 @@ impl yaml::K8sResource for ConfigMap {
         &self,
         _policy_mounts: &mut Vec<policy::KataMount>,
         _storages: &mut Vec<agent::Storage>,
+        _persistent_volume_claims: &Vec<PersistentVolumeClaim>,
```
I prefer the older convention, that shows more clearly which *.rs file a struct name is coming from. 

I would use pvc.rs and pvc::PersistentVolumeClaim instead of just PersistentVolumeClaim.

sprt replied:
Gotcha - any objection to renaming `persistent_volume_claim.rs` to `pvc.rs` then?

danmihai1 replied:
Please rename it that way.

sprt replied:
Done

ms-mahuber commented on src/tools/genpolicy/src/mount_and_storage.rs:
```diff
@@ -173,6 +177,63 @@ fn get_empty_dir_mount_and_storage(
+            fs_group: None,
+            source: "$(pci-device-id)".to_string(),
+            mount_point: source.to_string(),
+            fstype: "ext4".to_string(), // TODO: Don't hardcode this.
```
How hard is it to not hardcode this right now? Can we read this from the spec somehow?

sprt replied:
I think we should set this to `"$(fs-type)"` for now and not validate it (we're not validating this field for other storages). AFAIK extracting the FS type looks different for static vs. dynamic provisioning.

sprt replied:
Done above

sprt commented on src/tools/genpolicy/src/mount_and_storage.rs:
```diff
@@ -175,6 +177,63 @@ fn get_empty_dir_mount_and_storage(
+    let volume_pvc = yaml_volume.persistentVolumeClaim.as_ref().unwrap();
+    let is_blk_mount =
+        persistent_volume_claims.iter()
+        .find(|pvcResource| pvcResource.metadata.name.as_ref() == Some(&volume_pvc.claimName))
```
I think we should log a warning here if we can't find the PVC: either it wasn't provided via `-pvc` or it's not included in the spec file, but in either case we would not be able to determine if this is a block mount.

danmihai1 replied:
Warning sounds good - and perhaps return an Error value if we don't already.

sprt replied:
Done as a warning. Didn't error out as that would break existing users.

sprt commented:
Still TODO: add a `-pvc` flag.

danmihai1 commented:
> Still TODO: add a `-pvc` flag.

Let's try to reuse the older -c for all of ConfigMap, PVC, and in the future Secret input YAML files.

sprt commented:
Final iteration ready for review. Changes since last review:

* Addressed all comments above.
* Added the new `-a/--attached-resource-file` flag in a way that is backward-compatible and easy to cherry-pick upstream.
* Updated the genpolicy samples and tweaked the script to run in parallel.

ms-mahuber commented on src/tools/genpolicy/update_policy_samples.py:
```diff
@@ -1,3 +1,4 @@
+from concurrent.futures import ThreadPoolExecutor
```
Please assess whether parallel execution may have an effect on our layer cache.

danmihai1 replied:
Agreed - this is a good test case to detect possible cache corruptions. But:
- It might cost us some time, and
- If the cache gets corrupted, it might be hard to debug the corruption due to the random nature of parallel execution

So, I don't have a strong preference between parallel or serialized execution here.

ms-mahuber commented on src/tools/genpolicy/src/main.rs:
```diff
@@ -34,7 +36,11 @@ mod yaml;
 
 #[tokio::main]
 async fn main() {
+    if env::var("RUST_LOG").is_err() {
```
Please confirm with Dan whether we want to introduce this behavior.

sprt replied:
So the issue here is that warnings are not printed unless `RUST_LOG` is set to at least the `warn` level. So this code sets the env var accordingly, *unless* the user sets it. Thoughts?

danmihai1 replied:
Users having to enable warnings before they can see the warnings makes sense to me.

danmihai1 commented on src/tools/genpolicy/src/policy.rs:
```diff
@@ -396,16 +414,20 @@ impl AgentPolicy {
-                    secrets.push(secret);
-                }
+            // ConfigMap and Secret documents contain additional input for policy generation.
+            if kind.eq("ConfigMap") {
```
Does "make check" complain about not using match instead if/else/if/else ?

danmihai1 commented on src/tools/genpolicy/src/utils.rs:
```diff
@@ -97,6 +106,10 @@ impl Config {
     pub fn new() -> Self {
         let args = CommandLineOptions::parse();
 
+        if args.config_map_file.is_some() {
```
Let's keep the old -c and call all of these 3 types of files "configuration files" instead of "resource files".

Redent0r commented:
Please document new cmdline param in https://github.com/microsoft/kata-containers/blob/msft-main/src/tools/genpolicy/genpolicy-advanced-command-line-parameters.md

Redent0r commented on src/tools/genpolicy/src/mount_and_storage.rs:
```diff
@@ -113,7 +115,9 @@ pub fn get_mount_and_storage(
         get_empty_dir_mount_and_storage(settings, p_mounts, storages, yaml_mount, memory_medium);
-    } else if yaml_volume.persistentVolumeClaim.is_some() || yaml_volume.azureFile.is_some() {
+    } else if yaml_volume.persistentVolumeClaim.is_some() {
+        get_persistent_volume_claim_mount(settings, yaml_mount, yaml_volume, p_mounts, storages, persistent_volume_claims, "rprivate", "rw");
```
We might need to support "ro" access for pvc based on https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/readonlymany-disks#:~:text=You%20can%20now%20reference%20the%20new%20PersistentVolumeClaim%20in,mode%20to%20multiple%20nodes%20at%20the%20same%20time.

If so, may need to do something similar to https://github.com/microsoft/kata-containers/pull/175/files#diff-bfe7240cc8e3920d3200b19d567341adcfb828a69b91376faa7f6d48342b8295

If not, we may be able to remove `access` parameter from  `get_persistent_volume_claim_mount`, since it's always "rw"

sprt reviewed (commented):
New iteration - addressed all above comments.

sprt reviewed (commented):
Fixed last `make check` failure. Ready for review.

Redent0r reviewed (approved):
LGTM and I just verified it builds on Windows

sprt reviewed (commented):
Removed leftover TODO comment (related to make check) - merging now.
